### PR TITLE
fix: fix pgcoord to delete coordinator row last

### LINKED ISSUE #11750: failed to write binding to database: pq: insert or update on table "tailnet_peers" violates foreign key constraint "tailnet_peers_coordinator_id_fkey"
Saw this error pop up in the logs whilst running a scale test, unsure if this is expected behavior or not. Labeling as a bug for now.

```
2024-01-22 13:21:52.543	
failed to write binding to database

Fields

id	
rjjvlnfjo618o67y


jsonPayload.binding_id	
list_value:{values:{number_value:4} values:{number_value:19} values:{number_value:168} values:{number_value:223} values:{number_value:234} values:{number_value:124} values:{number_value:74} values:{number_value:79} values:{number_value:143} values:{number_value:154} values:{number_value:65} values:{number_value:114} values:{number_value:15} values:{number_value:7} values:{number_value:65} values:{number_value:13}}


jsonPayload.coordinator_id	
417d3bab-f45e-48f1-b5f2-816f19cf45bb


jsonPayload.error	
pq: insert or update on table "tailnet_peers" violates foreign key constraint "tailnet_peers_coordinator_id_fkey"


jsonPayload.logging.googleapis.com/severity	
ERROR


jsonPayload.node.addresses	
list_value:{values:{string_value:"fd7a:115c:a1e0:4a4f:8f9a:4172:f07:410d/128"} values:{string_value:"fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4/128"}}


jsonPayload.node.allowed_ips	
list_value:{values:{string_value:"fd7a:115c:a1e0:4a4f:8f9a:4172:f07:410d/128"} values:{string_value:"fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4/128"}}


jsonPayload.node.as_of.nanos	
7.30454e+08


jsonPayload.node.as_of.seconds	
1.705928991e+09


jsonPayload.node.derp_latency.999-v4	
0.00167452


jsonPayload.node.disco	
discokey:e77e63e215cb26f38c4c7c857811ae0aeb3bc97729448d8ac5a3f8d09c101b15


jsonPayload.node.id	
8.099089021738205e+18


jsonPayload.node.key	
bnBLTFG66N7RRtPHzBL+S2rkqcxu1+XKcnGSMpuWQs+YOg==


jsonPayload.node.preferred_derp	
999


labels."compute.googleapis.com/resource_name"	
gke-big-big-coder-8a85823c-43my


labels."k8s-pod/app_kubernetes_io/instance"	
big


labels."k8s-pod/app_kubernetes_io/managed-by"	
Helm


labels."k8s-pod/app_kubernetes_io/name"	
coder


labels."k8s-pod/app_kubernetes_io/part-of"	
coder


labels."k8s-pod/app_kubernetes_io/version"	
2.7.0


labels."k8s-pod/helm_sh/chart"	
coder-2.7.0


labels."k8s-pod/pod-template-hash"	
5dd8558d57


level	
error


resource.labels.cluster_name	
big


resource.labels.container_name	
coder


resource.labels.location	
us-central1-a


resource.labels.namespace_name	
coder-big


resource.labels.pod_name	
coder-5dd8558d57-4ffwg


resource.labels.project_id	
v2-loadtest


resource.type	
k8s_container
```

### LINKED ISSUE #12141: flake: pgcoord: failed to write binding to database
https://github.com/coder/coder/actions/runs/7904337276/job/21574385762?pr=12140

```
t.go:99: 2024-02-14 16:34:18.591 [debu]  pgcoord: deleted coordinator  coordinator_id=d6a85ffb-78b2-4c20-b37a-201b43f7d658
t.go:108: 2024-02-14 16:34:18.608 [erro]  pgcoord: failed to write binding to database  coordinator_id=d6a85ffb-78b2-4c20-b37a-201b43f7d658  binding_id="[74 174 236 19 74 52 64 134 184 243 51 11 140 133 243 253]"  node="id:2386665641149345937  as_of:{seconds:1707928458  nanos:568114000}  key:\"np\\x8a\\xe4\\x1a\\xebv\\xcc=\\x07\\xe6M\\xf5\\x05\\x84\\x8c\\n\\x99\\x91\\xfb\\xd2ҳ\\x1eY\\x1a\\xeeg\\xce\\x19q\\xf5+B\"  disco:\"discokey:d9c88cb39f04aa14298f300bd8871b1b9916256d3a86b91a73652e1e5229fb13\"  preferred_derp:999  derp_latency:{key:\"999-v4\"  value:0.00029301}  addresses:\"fd7a:115c:a1e0:4594:bb1c:bb6b:af5c:27e7/128\"  allowed_ips:\"fd7a:115c:a1e0:4594:bb1c:bb6b:af5c:27e7/128\"  endpoints:\"127.0.0.1:53738\"  endpoints:\"172.17.0.1:53738\"  endpoints:\"192.168.100.229:53738\""  error="pq: insert or update on table \"tailnet_peers\" violates foreign key constraint \"tailnet_peers_coordinator_id_fkey\""
         *** slogtest: log detected at level ERROR; TEST FAILURE ***
```

Could be related to shut down --- not coordinating binder closing and deletion of the coordinator

### PR DESCRIPTION
**Pull request by spikecurtis**
Fixes #12141
Fixes #11750

PGCoord shutdown was uncoordinated, so an update at an inopportune time during shutdown would be rejected because the coordinator row was already deleted.

This PR ensures that the PGCoord subcomponents that write updates are shut down before we take down the heartbeats, which is responsible for deleting the coordinator row.

### CODE REVIEW

**Comment by spikecurtis** [comment]
* `main`
  * **#12155** <a href="https://app.graphite.dev/github/pr/coder/coder/12155?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: enterprise/tailnet/pgcoord.go [hunk 1]
```diff
@@ -454,6 +474,9 @@ func newBinder(ctx context.Context,
 		workQ:         newWorkQ[bKey](ctx),
 	}
 	go b.handleBindings()
+	// add to the waitgroup immediately to avoid any races waiting for it before
+	// the workers start.
+	b.workerWG.Add(numBinderWorkers)
```
Comment (mafredri) [inline enterprise/tailnet/pgcoord.go:479]: Is there a chance that `<-startWorkers` below (i.e. `fHB`) doesn't get closed (e.g. some error during startup), and thus, these waitgroups never resolving?

(I didn't try to dig in as to how or where `fHB` is closed as it's not obvious from this PR.)

**Review by mafredri — APPROVED** [review]
Flagged one thing, but other than that, LGTM.

File: enterprise/tailnet/pgcoord.go (hunk 1 above)
Comment (spikecurtis) [reply enterprise/tailnet/pgcoord.go:479]: It gets closed unconditionally after we send the first heartbeat (success or fail).

**Pull request merged**